python/pyspark/sql/types.py

```diff
@@ -1119,8 +1120,8 @@ def Dict(d):
     class Row(tuple):
 
         """ Row in DataFrame """
-        __DATATYPE__ = dataType
-        __FIELDS__ = tuple(f.name for f in dataType.fields)
+        __datatype = dataType
```

Comment:
It looks like `Row` held a reference to its `DataType` even in the old code, so I guess the only new references here are the ones that we're adding to the functions / arrays / etc. returned from the other branches of `_create_cls`.  I suppose that we could set those objects' `__dataType__` fields inside of the `_create_cls` function instead of setting them in `_restore_object` if you think that would be easier to understand.  Not a huge deal to set it outside, though.